livemaker/cli/lmlsb.py

```diff
@@ -1159,3 +1159,154 @@ def scan_menus(lsb, csv_data, file_name, extract=False, patch=False):
+        append = False
+
+    with open(csv_file, ("a" if append else "w"), newline="\n") as csvfile:
+        csv_writer = csv.writer(csvfile, delimiter=",", quotechar='"', quoting=csv.QUOTE_MINIMAL)
+        if not append:
+            csv_writer.writerow(["Filename", "Command line", "Command subindex", "Original entry", "Translated entry"])
+        for row in csv_data:
+            csv_writer.writerow(row)
+
+    print("{} Text lines extracted.".format(len(csv_data)))
+
+
+@lmlsb.command()
+@click.argument("lsb_file", required=True, type=click.Path(exists=True, dir_okay="False"))
+@click.argument("csv_file", required=True, type=click.Path(exists=False, dir_okay="False"))
+@click.option("--no-backup", is_flag=True, default=False, help="Do not generate backup of original lsb file.")
+def inserttext(lsb_file, csv_file, no_backup):
+    """Apply translated texts from the given csv file on a given lsb file.
+    The csv file have to be created by the extracttext function and translations have to be added.
+    The original LSB file will be backed up to <lsb_file>.bak unless the
+    --no-backup option is specified.
+    """
+    print("Patching {} ...".format(lsb_file))
+
+    with open(lsb_file, "rb") as f:
+        data = f.read()
+    try:
+        lsb = LMScript.from_lsb(data)
+    except BadLsbError as e:
+        sys.exit("Failed to parse file: {}".format(e))
+
+    csv_data = []
+
+    with open(csv_file, newline="\n") as csvfile:
+        csv_reader = csv.reader(csvfile, delimiter=",", quotechar='"')
+        for row in csv_reader:
+            csv_data.append(row)
+
+    patched = scan_texts(lsb, csv_data, lsb_file, patch=True)
+
+    if not patched:
+        sys.exit("No text lines patched, so not need to change LSB file.")
+
+    if not no_backup:
+        print("Backing up original LSB.")
+        shutil.copyfile(str(lsb_file), "{}.bak".format(str(lsb_file)))
+    try:
+        new_lsb_data = lsb.to_lsb()
+        with open(lsb_file, "wb") as f:
+            f.write(new_lsb_data)
+        print("Wrote new LSB.")
+    except LiveMakerException as e:
+        sys.exit('Could not generate new LSB file: {}'.format(e))
+
+
+def scan_texts(lsb, csv_data, file_name, extract=False, patch=False):
+    """Scan given lsb object for TextIns, extract or patch TWdChar blocks
+    lsb: the lsb object loaded by LMScript.from_lsb
+    csv_data: translation data from the csv file
+    file_name: name of the lsb file
+    extract: true if you want to extract text
+    patch: true if you want to patch translated text
+    """
+    patched = False
+    for cmd_nr, cmd in enumerate(lsb.commands):
+        if cmd.type == CommandType.TextIns:
+            scenario = cmd.get('Text')
+            text_line = None
+            block_nr = 0
+            new_body = []
+            last_ch = None
+            cmd_patched = False
+            for w in scenario.body:
```

Comment:
for future reference, there's an api function for iterating over scripts in an LSB, which could save a few levels of indentation here

https://pylivemaker.readthedocs.io/en/latest/livemaker.lsb.html#livemaker.lsb.lmscript.LMScript.text_scenarios

Reply:
Thanks for hint to the API documentation. Stupid me have used the source until now...
Python is such a terrible language for analysing foreign code. The more i work with it the more i dislike python.